clrtrust-helper.c

```diff
@@ -0,0 +1,148 @@
+
+/* This is a helper application which is meant to be used with clrtrust, the
+ * Clear Linux Trust Store management tool. It is created for performance
+ * reasons, to process certificates in bulk as opposed to running openssl for
+ * each certificate file.
+ *
+ * It reads the list of files, one filename per line, from the standard input
+ * and produces output in form of: <filename>\t<fingerprint or hash>.
+ *
+ * Two modes are supported. If '-f' switch is specified on the command line, for
+ * each file a SHA-1 fingerprint is calculated. If '-s' is specified, then
+ * subject hash is produced.
+ */
+
+#include <openssl/bio.h>
+#include <openssl/pem.h>
+#include <openssl/x509.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+typedef enum {
+    MODE_FINGER = 1,    /* output fingerprint */
+    MODE_HASH,          /* output subject hash */
+    MODE_INVALID = 100
+} runmode_t;
+
+void print_help() {
+    char *help = "clrtrust-helper [-f|-s]\n"
+                 "Helper utility for clrtrust. Reads the list of files from"
+                 " stdin, one per line, and produces either fingerprint or"
+                 " subject hash for each file on stdout in form:\n"
+                 "    <filename><TAB><fingerprint or subject hash>\n";
+    puts(help);
+}
+
+int main(int argc, char **argv) {
+
+    runmode_t runmode = MODE_INVALID; /* global switch: whether to produce fingerprint
+                                   or subject hash. */
+
+    BIO *inbio = NULL;
+
+    /* fingerprint-related variables */
+    const EVP_MD *finger_type = NULL;
+    unsigned int finger_sz;
+    unsigned char finger[EVP_MAX_MD_SIZE];
+
+    /* subjecthash-related variables */
+    unsigned long subject_hash;
+
+    unsigned int i;
+    char c;
+
+    char *fname = NULL;
+    size_t sz = 0;
+
+    while ((c = getopt(argc, argv, "fs")) != -1) {
+        switch (c) {
+            case 'f':
+                if (runmode == MODE_INVALID) {
+                    runmode = MODE_FINGER;
+                } else {
+                    print_help();
+                    return 1;
+                }
+                break;
+            case 's':
+                if (runmode == MODE_INVALID) {
+                    runmode = MODE_HASH;
+                } else {
+                    print_help();
+                    return 1;
+                }
+                break;
+            default:
+                printf("Unrecognized option -%c.\n", c);
+                print_help();
+                return 1;
+        }
+    }
+    
+    if (runmode == MODE_INVALID) {
+        print_help();
+        return 1;
+    }
+
```

Comment:
I suggest below allowing the filenames to be specified on the command line.

If you don't then perhaps a check that nothing extra has been passed on the command line?

Maybe a '-z' mode to read the names NUL deliminated rather than newline deliminated?